lib/nrf_inbuilt_key/nrf_inbuilt_key.c

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <at_cmd.h>
+#include <at_cmd_parser/at_cmd_parser.h>
+#include <at_cmd_parser/at_params.h>
+#include <bsd_limits.h>
+#include <nrf_inbuilt_key.h>
+#include <stdio.h>
+#include <string.h>
+#include <zephyr.h>
+
+#define NRF_INBUILT_KEY_OP_LS "AT%CMNG=1"
+#define NRF_INBUILT_KEY_OP_RD "AT%CMNG=2"
+#define NRF_INBUILT_KEY_OP_WR "AT%CMNG=0"
+#define NRF_INBUILT_KEY_OP_RM "AT%CMNG=3"
+
+#define NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 (10)
+#define NRF_INBUILT_KEY_CR_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_LF_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_QUOTES_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE (1)
+#define NRF_INBUILT_KEY_COMMA_SIZE (1)
+#define NRF_INBUILT_KEY_INVALID_HANDLE (-1)
+
+static u8_t scratch_buf[4096];
+static char at_cmee_strings[2][10] = {"AT+CMEE=0", "AT+CMEE=1"};
+
+static int cmee_active(void)
+{
+	char response[sizeof("+CMEE: X\r\n") + 1];
+
+	memset(response, 0, sizeof(response));
+
+	enum at_cmd_state state;
+
+	int retval =
+		at_cmd_write("AT+CMEE?", response, sizeof(response), &state);
+	if (retval == 0) {
+		char *deactivated = strchr(&response[6], '0');
+
+		if (deactivated != NULL) {
+			return 0;
+		}
+
+		char *activated = strchr(&response[6], '1');
+
+		if (activated != NULL) {
+			return 1;
+		}
+	}
+
+	return -EIO;
+}
+
+static int cmee_set(bool enable)
+{
+	char response[10];
+	enum at_cmd_state state;
+
+	int retval = at_cmd_write(at_cmee_strings[(int)enable], response,
+				  sizeof(response), &state);
+
+	if (retval < 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int translate_error(int err)
+{
+	switch (err) {
+	case 513:
+		return -ENOENT;
+	case 514:
+		return -EPERM;
+	case 515:
+		return -ENOMEM;
+	case 518:
+		return -EACCES;
+	default:
+		return -EFAULT;
+	}
+}
+
+int nrf_inbuilt_key_write(nrf_sec_tag_t sec_tag,
+			  nrf_key_mgnt_cred_type_t cred_type, u8_t *p_buffer,
+			  u16_t buffer_len)
+{
+	if ((p_buffer == NULL) || (buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	u16_t scratch_buf_len = sizeof(NRF_INBUILT_KEY_OP_WR) + buffer_len +
+				(NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 * 2) +
+				(NRF_INBUILT_KEY_COMMA_SIZE * 3) +
+				(NRF_INBUILT_KEY_QUOTES_CHAR_SIZE * 2) +
+				NRF_INBUILT_KEY_CR_CHAR_SIZE +
+				NRF_INBUILT_KEY_LF_CHAR_SIZE +
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	u8_t *p_cmd = &scratch_buf[0];
```

Comment:
What about using char * as the type so we don't need to cast so much?

We could memset scratch_buf.

Reply:
Updated this line, and also changed the scratch buffer to char.